python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1112 @@
+        with nogil:
+            result = self.discovery.Inspect()
+        return pyarrow_wrap_schema(GetResultValue(result))
+
+    def finish(self):
+        cdef CResult[shared_ptr[CDataSource]] result
+        with nogil:
+            result = self.discovery.Finish()
+        return DataSource.wrap(GetResultValue(result))
+
+
+cdef class FileSystemDataSourceDiscovery(DataSourceDiscovery):
+
+    cdef:
+        CFileSystemDataSourceDiscovery* filesystem_discovery
+
+    def __init__(self, FileSystem filesystem not None, paths_or_selector,
+                 FileFormat format not None,
+                 FileSystemDiscoveryOptions options=None):
+        cdef:
+            FileStats file_stats
+            vector[CFileStats] stats
+            CResult[shared_ptr[CDataSourceDiscovery]] result
+
+        options = options or FileSystemDiscoveryOptions()
+        for file_stats in filesystem.get_target_stats(paths_or_selector):
+            stats.push_back(file_stats.unwrap())
+
+        result = CFileSystemDataSourceDiscovery.MakeFromFileStats(
+            filesystem.unwrap(),
+            stats,
+            format.unwrap(),
+            options.unwrap()
+        )
+        self.init(GetResultValue(result))
+
+    cdef init(self, shared_ptr[CDataSourceDiscovery]& sp):
+        DataSourceDiscovery.init(self, sp)
+        self.filesystem_discovery = <CFileSystemDataSourceDiscovery*> sp.get()
+
+
+cdef class DataSource:
+    """Basic component of a Dataset which yields zero or more data fragments.
+
+    A DataSource acts as a discovery mechanism of data fragments and
+    partitions, e.g. files deeply nested in a directory.
+    """
+
+    cdef:
+        shared_ptr[CDataSource] wrapped
+        CDataSource* source
+
+    def __init__(self):
+        _forbid_instantiation(self.__class__)
+
+    cdef void init(self, const shared_ptr[CDataSource]& sp):
+        self.wrapped = sp
+        self.source = sp.get()
+
+    @staticmethod
+    cdef wrap(shared_ptr[CDataSource]& sp):
+        cdef DataSource self
+
+        typ = frombytes(sp.get().type())
+        if typ == 'tree_data_source':
+            self = TreeDataSource.__new__(TreeDataSource)
+        elif typ == 'simple_data_source':
+            self = SimpleDataSource.__new__(SimpleDataSource)
+        elif typ == 'filesystem_data_source':
+            self = FileSystemDataSource.__new__(FileSystemDataSource)
+        else:
+            raise TypeError(typ)
+
+        self.init(sp)
+        return self
+
+    cdef shared_ptr[CDataSource] unwrap(self):
+        return self.wrapped
+
+    @property
+    def partition_expression(self):
+        cdef shared_ptr[CExpression] expression
+        expression = self.source.partition_expression()
+        if expression.get() == nullptr:
+            return None
+        else:
+            return Expression.wrap(expression)
+
+
+# TODO(kszucs) Rename it to InMemoryDataSource
+cdef class SimpleDataSource(DataSource):
```

Comment:
We should have a jira about introducing InMemoryDataSource, if I recall correctly we already have one?